File: mobile/lib/models/community.dart
import 'dart:convert';

class CommunityPost {
  final String id;
  final String chapterId;
  final String content;
  final String? imageUrl;
  final bool isPinned;
  final DateTime createdAt;
  final PostAuthor author;
  final int likesCount;
  final int commentsCount;
  final bool isLikedByMe;

  // Phase 2 Economic Engine fields
  final String postType; // general, lead, rfp
  final String visibility; // chapter, network, club
  final String? leadStatus; // open, in_progress, closed_won, closed_lost
  final String? budgetRange;
  final DateTime? deadline;
  final String? targetClubId;
  final String? targetClubName;
  final String? targetIndustryId;
  final String? targetIndustryName;
  final double? businessValue;

  CommunityPost({
    required this.id,
    required this.chapterId,
    required this.content,
    this.imageUrl,
    required this.isPinned,
    required this.createdAt,
    required this.author,
    required this.likesCount,
    required this.commentsCount,
    required this.isLikedByMe,
    this.postType = 'general',
    this.visibility = 'chapter',
    this.leadStatus,
    this.budgetRange,
    this.deadline,
    this.targetClubId,
    this.targetClubName,
    this.targetIndustryId,
    this.targetIndustryName,
    this.businessValue,
  });

  factory CommunityPost.fromJson(Map<String, dynamic> json) {
    return CommunityPost(
      id: json['id'],
      chapterId: json['chapter_id'],
      content: json['content'],
      imageUrl: json['image_url'],
      isPinned: json['is_pinned'],
      createdAt: DateTime.parse(json['created_at']),
      author: PostAuthor.fromJson(json['author']),
      likesCount: json['likes_count'],
      commentsCount: json['comments_count'],
      isLikedByMe: json['is_liked_by_me'],
      postType: json['post_type'] ?? 'general',
      visibility: json['visibility'] ?? 'chapter',
      leadStatus: json['lead_status'],
      budgetRange: json['budget_range'],
      deadline: json['deadline'] != null ? DateTime.parse(json['deadline']) : null,
      targetClubId: json['target_club_id'],
      targetClubName: json['target_club_name'],
      targetIndustryId: json['target_industry_id'],
      targetIndustryName: json['target_industry_name'],
      businessValue: json['business_value'] != null ? (json['business_value'] as num).toDouble() : null,
    );
  }
}

class PostAuthor {
  final String id;
  final String fullName;
  final String? profilePhoto;
  final String role;

  PostAuthor({
    required this.id,
    required this.fullName,
    this.profilePhoto,
    required this.role,
  });

  factory PostAuthor.fromJson(Map<String, dynamic> json) {
    return PostAuthor(
      id: json['id'],
      fullName: json['full_name'],
      profilePhoto: json['profile_photo'],
      role: json['role'],
    );
  }
}

class PostComment {
  final String id;
  final String postId;
  final String content;
  final DateTime createdAt;
  final PostAuthor author;

  PostComment({
    required this.id,
    required this.postId,
    required this.content,
    required this.createdAt,
    required this.author,
  });

  factory PostComment.fromJson(Map<String, dynamic> json) {
    return PostComment(
      id: json['id'],
      postId: json['post_id'],
      content: json['content'],
      createdAt: DateTime.parse(json['created_at']),
      author: PostAuthor.fromJson(json['author']),
    );
  }
}

File: backend/app/models/community.py
"""
Prime Business Network – Community Models.

Handles Chapter-scoped social feed including posts, likes, and comments.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import List

import enum
from sqlalchemy import Boolean, ForeignKey, String, Text, UniqueConstraint, Enum, Numeric, DateTime
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin


class PostType(str, enum.Enum):
    GENERAL = "general"
    LEAD = "lead"
    RFP = "rfp"


class PostVisibility(str, enum.Enum):
    CHAPTER = "chapter"
    NETWORK = "network"
    CLUB = "club"


class LeadStatus(str, enum.Enum):
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    CLOSED_WON = "closed_won"
    CLOSED_LOST = "closed_lost"


class CommunityPost(Base, TimestampMixin):
    __tablename__ = "community_posts"

    chapter_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("chapters.id", ondelete="CASCADE"), nullable=False, index=True
    )
    author_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    content: Mapped[str] = mapped_column(Text, nullable=False)
    image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    is_pinned: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    # Economic Engine Fields
    post_type: Mapped[PostType] = mapped_column(
        Enum(PostType), default=PostType.GENERAL, nullable=False, index=True
    )
    visibility: Mapped[PostVisibility] = mapped_column(
        Enum(PostVisibility), default=PostVisibility.CHAPTER, nullable=False, index=True
    )
    lead_status: Mapped[LeadStatus | None] = mapped_column(
        Enum(LeadStatus), nullable=True, index=True
    )
    
    target_club_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("horizontal_clubs.id", ondelete="SET NULL"), nullable=True
    )
    target_industry_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("industry_categories.id", ondelete="SET NULL"), nullable=True
    )
    
    budget_range: Mapped[str | None] = mapped_column(String(100), nullable=True)
    deadline: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    
    # Thank You For Business (TYFB) Value
    business_value: Mapped[float | None] = mapped_column(Numeric(15, 2), nullable=True)

    # Relationships
    author = relationship("User", lazy="joined")
    target_club = relationship("HorizontalClub", lazy="joined")
    target_industry = relationship("IndustryCategory", lazy="joined")
    likes = relationship("PostLike", back_populates="post", cascade="all, delete-orphan")
    comments = relationship("PostComment", back_populates="post", cascade="all, delete-orphan", order_by="PostComment.created_at")

    def __repr__(self) -> str:
        return f"<CommunityPost id={self.id} author={self.author_id}>"


class PostLike(Base, TimestampMixin):
    __tablename__ = "post_likes"

    post_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("community_posts.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    # Relationships
    post = relationship("CommunityPost", back_populates="likes")

    __table_args__ = (
        UniqueConstraint("post_id", "user_id", name="unique_user_post_like"),
    )


class PostComment(Base, TimestampMixin):
    __tablename__ = "post_comments"

    post_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("community_posts.id", ondelete="CASCADE"), nullable=False, index=True
    )
    author_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    content: Mapped[str] = mapped_column(Text, nullable=False)

    # Relationships
    post = relationship("CommunityPost", back_populates="comments")
    author = relationship("User", lazy="joined")

    def __repr__(self) -> str:
        return f"<PostComment id={self.id} author={self.author_id}>"
